sos/plugins/kubernetes.py

```diff
@@ -15,40 +15,130 @@
 # Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 
 from sos.plugins import Plugin, RedHatPlugin
+from os import path
 
 
 class kubernetes(Plugin, RedHatPlugin):
 
     """Kubernetes plugin
     """
 
-    option_list = [("podslog", "capture logs for pods", 'slow', False)]
+    # Red Hat Atomic Platform and OpenShift Enterprise use the
+    # atomic-openshift-master package to provide kubernetes
+    packages = ('kubernetes', 'atomic-openshift-master')
+    files = ("/etc/origin/master/master-config.yaml",)
+
+    option_list = [
+        ("all", "use --all-namespaces for all kubectl commands",
+            'fast', True),
+        ("describe", "capture descriptions of all kube resources",
+            'fast', True),
+        ("podlogs", "capture logs for pods", 'slow', False),
+    ]
+
+    def check_is_master(self):
+        if any([path.exists("/var/run/kubernetes/apiserver.key"),
+                path.exists("/etc/origin/master/master-config.yaml")
+                ]):
+            return True
+        else:
+            return False
 
     def setup(self):
         self.add_copy_spec("/etc/kubernetes")
         self.add_copy_spec("/var/run/flannel")
 
-        # Kubernetes master info
-        self.add_cmd_output("kubectl version")
-        self.add_cmd_output("kubectl get -o json pods")
-        self.add_cmd_output("kubectl get -o json nodes")
-        self.add_cmd_output("kubectl get -o json services")
-        self.add_cmd_output("kubectl get -o json replicationController")
-        self.add_cmd_output("kubectl get -o json events")
-        self.add_cmd_output("journalctl -u kubelet")
-        self.add_cmd_output("journalctl -u kube-apiserver")
-        self.add_cmd_output("journalctl -u kube-controller-manager")
-        self.add_cmd_output("journalctl -u kube-scheduler")
-        self.add_cmd_output("journalctl -u kube-proxy")
-
-        if self.get_option('podslog'):
-            result = self.get_command_output("kubectl get pods")
-            if result['status'] == 0:
-                for line in result['output'].splitlines()[1:]:
-                    pod_name = line.split(" ")[0]
-                    self.add_cmd_output([
-                        "{0} log {1}".format("kubectl", pod_name)
-                    ])
+        svcs = [
+            'kubelet', 'kube-apiserver', 'kube-proxy', 'kube-scheduler',
+            'kube-controller-manager'
+        ]
+
+        for svc in svcs:
+            self.add_cmd_output("journalctl -u {}".format(svc))
+
+        # We can only grab kubectl output from the master
+        if self.check_is_master():
+            kube_bin = "kubectl "
+            if path.exists('/etc/origin/master/admin.kubeconfig'):
+                kube_bin += "--config=/etc/origin/master/admin.kubeconfig"
```

Comment:
This isn't a coding style question. It's just a case of following conventions - in other words, don't invent something new if there are existing things doing the same kind of thing. If it's likely that something is already being done elsewhere, go and look at it before you do your thing (surely you do that anyway to see if there are existing APIs or other abstractions you can take advantage of??).

Unsurprisingly building and running command strings is pretty common in sos.

It is easier on everyone if they use similar constructs.